src/date.py

```diff
@@ -26,20 +26,34 @@ def build(self, date_object):
         return datetime(date_object["year"], date_object["month"], date_object["day"],
                         date_object["hour"], date_object["minute"], date_object["second"])
 
-    def from_exif(self, exif, user_regex=None):
+    def from_exif(self, exif, timestamp=False, user_regex=None):
         keys = ['SubSecCreateDate', 'SubSecDateTimeOriginal', 'CreateDate', 'DateTimeOriginal']
 
         datestr = None
+        parsed_date = None
 
         for key in keys:
             if key in exif:
                 datestr = exif[key]
                 break
+        
+        
+        if isinstance(datestr,str): #sometimes this returns an int
+            # sometimes exif data can return all zeros
+            # check to see if valid date first
+            if datestr:
+                parsed_date = self.from_datestring(datestr)
+        else:
+            parsed_date = {'date': None, 'subseconds': ''}
 
-        if datestr:
-            return self.from_datestring(datestr)
+        if parsed_date.get("date") is not None:
+            return parsed_date
         else:
-            return self.from_filename(user_regex)
+            if timestamp: return self.from_timestamp()
```

Comment:
I don't like that the timestamp takes precedence over filename regex. The case when the date will not be in EXIF and in the filename is rare. Currently if you pass the timestamp option it will ignore the filename date but in reality the timestamp must be last chance to get the date as it is possible to be totally incorrect date.

The perfect case is to try to get date from filename and if it fails then to try the timestamp, if enabled.

Reply:
yeah, you're right - it didn't matter in my case since i was using this on old files where the camera didn't set the date in the filename (not like it had the date set correctly anyway, if only i had done that each time!) but this really should be an option of last resort (which nevertheless was useful for me in a lot of files), i will juggle those things around